core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -620,11 +620,13 @@ private[spark] class BlockManager(
    * Note that this method must be called without any BlockInfo locks held.
    */
   def reregister(): Unit = {
-    // TODO: We might need to rate limit re-registering.
-    logInfo(s"BlockManager $blockManagerId re-registering with master")
-    master.registerBlockManager(blockManagerId, diskBlockManager.localDirsString, maxOnHeapMemory,
-      maxOffHeapMemory, storageEndpoint)
-    reportAllBlocks()
+    if (!SparkEnv.get.isStopped) {
```

Comment:
Hmm...so, let's assume such a scenario: driver command a shutdown request to the executor. Before the executor receives the shutdown request, it finds that it itself has been removed from the driver when reporting a block to the driver. Then, the executor starts to reregister itself. Since the executor hasn't received shutdown request, which means `SparkEnv.get.isStopped=false`, it successfully reregisters with the driver. Soon, the executor receives the shutdown request and exits itself. 

So, this change doesn't fix the issue thoroughly, right? @wankunde

Reply:
@Ngone51 Thanks for you review.

Yes,  this PR can not fix the issue above,  but I also think  that adding `!SparkEnv.get.isStopped` constraint is helpful as I have found several executors re-register when they are shutting down by driver.

I very agree to fix this issue in `HeartbeatReceiver` and this PR can be closed.

Reply:
You can either update in this PR or in a separate PR as you like.

Reply:
I have updated the PR description.

Reply:
Sorry, I mean you can update the fix with `HeartbeatReceiver` in this PR.